OAuth 2 Provider for frappe

revant opened the pull request:
brought working oauth 2 provider from mnt_oauth into frappe

revant committed:
OAuth 2.0 Classes Added

Added oauthlib in requirement.txt
Added class WebApplicationServer for Authorization code grant and Bearer tokens.
Added class OAuthWebRequestValidator for Validating Oauth Request for Web Application

revant committed:
copied code from mnt_oauth needs review

Commit:
[New] OAuth2 Doctypes and endpoints added

Integrations > OAuth Provider Settings
Integration Broker > OAuth Bearer Token
Integration Broker > OAuth Authorization Code
Integration Broker > OAuth Client
endpoints for authorize, approve, get_token, revoke_token and test_resource

vjFaLk commented on frappe/integration_broker/doctype/oauth_bearer_token/oauth_bearer_token.py:
```diff
@@ -0,0 +1,13 @@
+class OAuthBearerToken(Document):
+	def validate(self):
+		if not self.expiration_time:
+	 		self.expiration_time = frappe.utils.datetime.datetime.strptime(self.creation, "%Y-%m-%d %H:%M:%S.%f") + frappe.utils.datetime.timedelta(seconds=self.expires_in)
```
Maybe import the functions instead of using them inline for the sake of cleanliness?

vjFaLk commented on frappe/integration_broker/oauth2.py:
```diff
@@ -0,0 +1,139 @@
+
+			scopes, credentials = oauth_server.validate_authorization_request(uri, http_method, body, headers)
+
+			skipauth = frappe.db.get_value("OAuth Client", credentials['client_id'], "skip_authorization")
```
Maybe skip_auth instead of skipauth. 

Skipauth sounds like a nice band name though.

revant committed:
oauth2.py: renamed skipauth to skip_auth

revant commented:
Endpoints available in frappe/integration_broker/oauth2.py

![diagram1](https://cloud.githubusercontent.com/assets/1040161/19417391/48a2620e-93c9-11e6-83bd-fed5494e8193.png)

get_token endpoint is used for 1) getting access_token from auth code and 2) auth code from refresh token

there is one more endpoint revoke_token to revoke access_tokens

gaurav-naik commented:
## Authorization Code Endpoint

### Authorization Request
URL:
   
```
0.0.0.0:8000/api/method/frappe.integration_broker.oauth2.authorize
```
Params:
```
client_id = <client ID of registered app>
scope = <access scope, e.g. scope=project will allow you to access project doctypes.>
response_type = "code"
redirect_uri = <redirect uri from OAuth Client>
```
  
### Confirmation Dialog
Click 'Allow'  to receive authorization code in redirect uri.

```
http://localhost:3000/oauth_code?code=plkj2mqDLwaLJAgDBAkyR1W8Co08Ud
```

## Get Token Endpoint

### Get Access Token
URL:
```
0.0.0.0:8000/api/method/frappe.integration_broker.oauth2.get_token
```
Params:
```
grant_type = "authorization_code"
code = <code received in redirect uri after confirmation>
redirect_uri= <valid redirect uri>
client_id = <client ID of app from OAuth Client>
```
Response:
```
{
	"message": {
	"access_token": "pNO2DpTMHTcFHYUXwzs74k6idQBmnI",
	"token_type": "Bearer",
	"expires_in": 3600,
	"refresh_token": "cp74cxbbDgaxFuUZ8Usc7egYlhKbH1",
	"scope": "project"
	}
}
```
### Refresh Token
URL:
```
0.0.0.0:8000/api/method/frappe.integration_broker.oauth2.get_token
```
Params:
```
grant_type = "refresh_token"
refresh_token = <refresh token from the response of get_token call with grant_type=authorization_code>
redirect_uri= <valid redirect uri>
client_id = <client ID of app from OAuth Client>
```
Response:
```
{
  "message": {
    "access_token": "Ywz1iNk0b21iAmjWAYnFWT4CuudHD5",
    "token_type": "Bearer",
    "expires_in": 3600,
    "refresh_token": "PNux3Q8Citr3s9rl2zEsKuU1l8bSN5",
    "scope": "project"
  }
}
```

## Revoke Token Endpoint
URL:
```
0.0.0.0:8000/api/method/frappe.integration_broker.oauth2.revoke_token
```
Params:
```
client_id = <client ID of app from OAuth Client>	
token = <access token to be revoked>
```
Response:
```
{"status": 200}
```

rmehta commented on frappe/integration_broker/oauth2.py:
```diff
@@ -0,0 +1,139 @@
+from urllib import quote, urlencode
+from frappe.integrations.doctype.oauth_provider_settings.oauth_provider_settings import get_oauth_settings
+
+#Variables required across requests
```
A good place to add some docs:

See

https://github.com/frappe/frappe/blob/develop/frappe/integrations/doctype/razorpay_settings/razorpay_settings.py#L12

rmehta commented:
@revant @gaurav-naik Good start :)

1. Add screenshots of the doctypes and success / failure pages in the PR
1. Add a new article in the `docs` explaining a new dev on how to integrate this.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' of https://github.com/revant/frappe into develop

Commit:
Merge branch 'develop' of https://github.com/frappe/frappe into develop

Commit:
Merge branch 'develop' of https://github.com/frappe/frappe into develop

revant committed:
Merge branch 'develop' of https://github.com/frappe/frappe into develop

revant committed:
[Fix] Parse URL instead of storing it in settings

Commit:
Merge branch 'develop' of https://github.com/revant/frappe into develop

revant commented:
## Screenshots of DocTypes

### OAuth Provider Settings
![oauthprovidersettings](https://cloud.githubusercontent.com/assets/1040161/19489963/de88a6f8-958a-11e6-8d1e-faa6b4f30122.png)
<hr>
### OAuth Authorization Code
![oauthauthorizationcode](https://cloud.githubusercontent.com/assets/1040161/19489989/f73f1420-958a-11e6-8278-e7962d8df695.png)
<hr>
### OAuth Bearer Token
![oauthbearertoken](https://cloud.githubusercontent.com/assets/1040161/19490005/00c43444-958b-11e6-8747-8771761a95b7.png)
<hr>
### OAuth Client
![oauthclient](https://cloud.githubusercontent.com/assets/1040161/19490021/0a7b2218-958b-11e6-87b6-411feb359858.png)
<hr>

revant committed:
[Fix] get skip_auth from OAuth Provider Settings

Commit:
Merge branch 'develop' of https://github.com/revant/frappe into develop

Commit:
Success URL format. Failure URL added. Confirmation dialog layout changed.

rmehta commented:
@revant @gaurav-naik 

Thanks! Please add some basic docs. Add a new page in `frappe/docs/user/en/guides/integration`, then we are good to go!

You can add
- [ ] How to setup oauth
- [ ] How to test
- [ ] Screeshots

revant commented:
At the moment we're trying to merge authorized user derived from valid access_token into frappe's role and permission system using frappe/frappe/api.py
so that following also starts working using the existing user role permission system.

```
/api/method/dotted.path.to.function?param1=aaa&param2=bbb&access_token=valid_access_token 
/api/resource/DocType/DocName?fields=["name","description"]&access_token=valid_access_token 
```
and other frappe's built in REST API with additional parameter of access_token.

Should this be priority? any tip or suggestion?

Right now the permission system is handled with whitelisted function using the reference code mentioned in oauthlib documentation.
https://github.com/revant/frappe/blob/bc2cb9a5c1e76933b60d810e46e3d933559c079e/frappe/integration_broker/oauth2.py#L125

We'll also push Documentation mentioned in previous comment. Working on it.

revant committed:
Merge branch 'develop' of https://github.com/frappe/frappe into develop

rmehta commented:
Move the access_token system to another pull-request.

Did you finish the docs?

revant committed:
Validate access token if passed during use of REST API

revant commented:
I'll push docs today or tomorrow morning

On 19-Oct-2016 9:36 pm, "Rushabh Mehta" <notifications@github.com> wrote:

> Move the access_token system to another pull-request.
>
> Did you finish the docs?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/frappe/frappe/pull/2167#issuecomment-254860274>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA_fIfuP-CcFqmWLJiaEhAv5qfp4_edVks5q1kAggaJpZM4KX8X_>
> .
>

Commit:
OAuth Confirmation colours fixed

gaurav-naik commented:
### OAuth Confirmation Dialog (Updated)

![oauth_conf_colors](https://cloud.githubusercontent.com/assets/17795124/19531316/e4cad16a-9654-11e6-9335-88cc1b05c4e6.png)

Commit:
Multiple Changes

Added links for OAuth under Integrations in Module list.
Updated permissions on OAuth doctypes.
Updated layout of OAuth Client doctype.

revant committed:
[Docs] Integrations > How to setup OAuth

revant committed:
[Docs] Integration > Using OAuth

revant committed:
Merge branch 'develop' of https://github.com/frappe/frappe into develop

revant committed:
[Fix] get_token endpoint must to handle POST request

revant committed:
[Fix] http verbs and responses for OAuth 2.0 Endpoints

revant committed:
[Fix] accept oauth2 access_token from headers

revant committed:
Removed unused imports from api.py

revant committed:
Merge branch 'develop' of https://github.com/frappe/frappe into develop

revant committed:
Merge branch 'develop' of https://github.com/frappe/frappe into develop

revant committed:
Merge branch 'develop' of https://github.com/frappe/frappe into develop

revant committed:
Merge branch 'develop' of https://github.com/frappe/frappe into develop

javierwonghf commented:
@revant or anyone whom might know, would you happen to know how does `frappe.request.url` get the request URL from? 

Reason why I ask is, I have the entire site served under https and I am having issues with OAuth2 during the authorization part, I get an error saying that there is Mixed-Content returned and after digging through the code I found that `frappe.request.url` shows the authorization request URL but in http and not https.

Thanks.
